client/types/kits/reader.go

```diff
@@ -0,0 +1,334 @@
+		return
+	}
+	var item Item
+	var hdr *tar.Header
+	tr := tar.NewReader(rp.rdr)
+	for {
+		if hdr, err = tr.Next(); err != nil {
+			if err == io.EOF {
+				break
+			}
+			return
+		}
+		nm := filepath.Base(hdr.Name)
+		if nm == ManifestName || nm == ManifestSigName {
+			continue
+		}
+		if item, err = rp.getItem(nm); err != nil {
+			return err
+		}
+		if err = cb(item.Name, item.Type, item.Hash, tr); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (rp *Reader) getItem(n string) (item Item, err error) {
+	for _, v := range rp.manifest.Items {
+		if v.Filename() == n {
+			item = v
+			return
+		}
+	}
+	err = fmt.Errorf("%s not found in manifest", n)
+	return
+}
+
+// SigVerificationFunc is the function type used to validate a manifest signature. It is
+// passed the manifest and signature as slices of bytes. For standard public-key signature
+// verification, use a lambda function which captures the key.
+type SigVerificationFunc func(manifest []byte, sig []byte) error
+
+// Verify reads a kit from the rdr and checks that all items are valid. If sigVerify is not
+// nil, it will be called to verify the manifest signature.
+func Verify(rdr io.Reader, sigVerify SigVerificationFunc) (signed bool, manifest Manifest, err error) {
+	fileHashes := map[string][sha256.Size]byte{}
+	var m, s []byte
+	var hdr *tar.Header
+
+	tr := tar.NewReader(rdr)
+	for {
+		if hdr, err = tr.Next(); err != nil {
+			if err == io.EOF {
+				break
+			}
+			return
+		}
+		nm := filepath.Base(hdr.Name)
+		if nm == ManifestName {
+			lr := io.LimitedReader{
+				R: tr,
+				N: maxManifestSize,
+			}
+			if m, err = ioutil.ReadAll(&lr); err != nil {
+				return
+			}
+		} else if nm == ManifestSigName {
+			lr := io.LimitedReader{
+				R: tr,
+				N: maxManifestSigSize,
+			}
+			if s, err = ioutil.ReadAll(&lr); err != nil {
+				return
+			}
+		} else {
+			//we are just hashing the file and parking it in the map
+			h := sha256.New()
+			io.Copy(h, tr)
+			x := h.Sum(nil)
+			var v [sha256.Size]byte
+			if len(x) != sha256.Size {
+				err = ErrInvalidHash
+				return
+			}
+			copy(v[0:sha256.Size], x)
+			fileHashes[hdr.Name] = v
+		}
+	}
+	//check that we got a manifest
+	if m == nil {
+		err = ErrMissingManifest
+		return
+	}
+	//sigs match, lets checkout the files and manifest
+	if err = manifest.Unmarshal(m); err != nil {
+		return
+	}
+	if len(manifest.Items) != len(fileHashes) {
+		err = fmt.Errorf("%w - Expected %d items, got %d", ErrManifestMismatch, len(manifest.Items), len(fileHashes))
+		return
+	}
+	for _, item := range manifest.Items {
+		if hsh, ok := fileHashes[item.Filename()]; !ok {
+			err = fmt.Errorf("%s (%s) in manifest missing from kit", item.Name, item.Filename())
+			return
+		} else if hsh != item.Hash {
+			err = fmt.Errorf("%s in kit is corrupted", item.Name)
+			return
+		}
+	}
+	// Do the signing verification if provided
+	if sigVerify != nil {
+		if err = sigVerify(m, s); err != nil {
+			err = ErrInvalidSignature
```

Comment:
seems like this should hand back the sigVerify error wrapped so that the user can still get access to their returned error: `err = fmt.Errorf("%w - %v", err, ErrInvalidSignature)`